### PR TITLE
fix: canvas recording patches

### PR DESCRIPTION
**Pull request by daibhin**
## Changes

Closes https://github.com/PostHog/posthog-flutter/issues/69

Main aim is to support session recording in Flutter!

Does so in two ways:

- Support canvas capture in shadow roots
  - Query selectors do not include shadow root elements by default
  - [Updates `getCanvas` method to also traverse shadow roots](https://github.com/PostHog/posthog-js/compare/dn-fix/canvas-patches?expand=1#diff-b3df0a7a5c244118dd91288bff1fcc2b6587e659c8eeee7cefd574b7829cffa8R1423-R1427)
  - Hat tip to an unmerged PR https://github.com/rrweb-io/rrweb/pull/1413

- Capture correct canvas size
  - Canvas elements accept `width` and `height` attributes which set the internal coordinates of the canvas
  - They cannot always be relied upon to reflect the rendered width / height of the canvas e.g. if style / class attributes are added (as is the case in the canvas Flutter renders) the canvas coordinate "size" might be different to the rendered viewport meaning the captured base64 data will look zoomed in / out
  - Use the `clientWidth` and `clientHeight` (based on [this description](https://stackoverflow.com/questions/21064101/understanding-offsetwidth-clientwidth-scrollwidth-and-height-respectively)) properties of the canvas to [resize the output image](https://github.com/PostHog/posthog-js/compare/dn-fix/canvas-patches?expand=1#diff-b3df0a7a5c244118dd91288bff1fcc2b6587e659c8eeee7cefd574b7829cffa8R1469-R1487) when calling `createImageBitmap`
  - Playback goes from this to this:

|Before|After|
|----|----|
| Screenshot 2024-03-14 at 17 06 25 | Screenshot 2024-03-14 at 17 06 08 |

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers
- [ ] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

### CODE REVIEW

**Review by pauldambra — APPROVED** [review]
nice

I ended up switching off auto-formatting and finding out how to save without formatting no matter what to avoid adding lots of whitespacee to the patch

but that's nitpicky

**Comment by marandaneto** [comment]
> nice
> 
> I ended up switching off auto-formatting and finding out how to save without formatting no matter what to avoid adding lots of whitespacee to the patch
> 
> but that's nitpicky

yeah that sounds like a good idea, its also easier to review/less chance of conflicts

**Comment by pauldambra** [comment]
@daibhin I did CMD + shift + P in vs code, found "save without formatting", and then bound some keys to it.

For me I bound CMD + S  since I don't work in vscode full time. But then we'll see the diff of just the lines you changed 

pretty plz and thank you 😊

File: patches/rrweb@2.0.0-alpha.11.patch [hunk 1]
```diff
@@ -1133,19 +1133,70 @@ index 81ad4520d1435d088752a85f9a55a68ad6e0f514..e19dec6cc98ea782d53328853ca1b46d
...
+@@ -111,6 +122,12 @@ class CanvasManager {
+                 const id = this.mirror.getId(canvas);
+                 if (snapshotInProgressMap.get(id))
+                     return;
 +
 +                // The browser throws if the canvas is 0 in size
 +                // Uncaught (in promise) DOMException: Failed to execute 'createImageBitmap' on 'Window': The source image width is 0.
 +                // Assuming the same happens with height
 +                if (canvas.width === 0 || canvas.height === 0) return;
++
```
Comment (daibhin) [inline patches/rrweb@2.0.0-alpha.11.patch:1176]: Moved this up based on https://github.com/rrweb-io/rrweb/pull/1422/commits/d860a4b5338385254d316ad5d82ac10f681de40d

**Comment by daibhin** [comment]
@pauldambra done! TIL disabling Prettier requires VS Code to be restarted to take effect

**Review by pauldambra — APPROVED** [review]
🚢 

way nicer diff 😍

**Pull request merged**